### ISSUE TITLE
Records with modifiers have `never` keys when put through `StaticDecode`.

### ISSUE THREAD

**Issue by HamishWHC** [sinclairzx81/typebox#555]
And another issue (sorry)... I've found a simpler example for this one though.

Example:
```ts
const Records = Type.Object({
	neither: Type.Record(Type.String(), Type.Object({})),
	optional: Type.Optional(Type.Record(Type.String(), Type.Object({}))),
	readonly: Type.Readonly(Type.Record(Type.String(), Type.Object({}))),
	optional_readonly: Type.Optional(Type.Readonly(Type.Record(Type.String(), Type.Object({})))),
});

type Records = StaticDecode<typeof Records>;
/*
type Records = {
    readonly optional_readonly?: Record<never, {}> | undefined;
    readonly readonly: Record<never, {}>;
    optional?: Record<never, {}> | undefined;
    neither: Record<string, {}>;
}
*/
```

**Comment by sinclairzx81** [comment]
@HamishWHC Hi,

Yeah, this one looks like a bug, however I probably won't have time to look into this till the end of the week. Will do some initial investigations in the interim.

Cheers
S

**Comment by sinclairzx81** [comment]
@HamishWHC Quick update

The following may be a potential fix for this issue, however this type is getting somewhat unwieldy. The fix here unwraps both readonly and optional into X and Y types specifically. This enables the Record `Key` type to infer correctly.

```typescript
// --------------------------------------------------------------------------
// TTransform
// --------------------------------------------------------------------------
// prettier-ignore
export type DecodeModifier<D extends TSchema, T extends TSchema> =
  T extends TReadonly<T> & TOptional<T> ? TReadonly<TOptional<AssertType<D>>> :
  T extends TReadonly<T> ? TReadonly<D> :
  T extends TOptional<T> ? TOptional<D> :
  D
// prettier-ignore
export type DecodeProperties<T extends TProperties> = {
  [K in keyof T]: DecodeType<T[K]>
}
// prettier-ignore
export type DecodeRest<T extends TSchema[]> = T extends [infer L, ...infer R]
  ? [DecodeType<AssertType<L>>, ...DecodeRest<AssertRest<R>>]
  : []
// prettier-ignore
export type DecodeType<T extends TSchema, X = ReadonlyUnwrapType<T>, Y = OptionalUnwrapType<AssertType<X>>> =
  Y extends TTransform<infer _, infer R> ? DecodeModifier<TUnsafe<R>, T> : 
  Y extends TArray<infer S> ? DecodeModifier<TArray<DecodeType<S>>, T> : 
  Y extends TAsyncIterator<infer S> ? DecodeModifier<TAsyncIterator<DecodeType<S>>, T> : 
  Y extends TConstructor<infer P, infer R> ? DecodeModifier<TConstructor<AssertRest<DecodeRest<P>>, DecodeType<R>>, T> : 
  Y extends TFunction<infer P, infer R> ? DecodeModifier<TFunction<AssertRest<DecodeRest<P>>, DecodeType<R>>, T> : 
  Y extends TIntersect<infer S> ? DecodeModifier<TIntersect<AssertRest<DecodeRest<S>>>, T> : 
  Y extends TIterator<infer S> ? DecodeModifier<TIterator<DecodeType<S>>, T> : 
  Y extends TNot<infer S> ? DecodeModifier<TNot<DecodeType<S>>, T> : 
  Y extends TObject<infer S> ? DecodeModifier<TObject<Evaluate<DecodeProperties<S>>>, T> : 
  Y extends TPromise<infer S> ? DecodeModifier<TPromise<DecodeType<S>>, T> : 
  Y extends TRecord<infer K, infer S> ? DecodeModifier<TRecord<K, DecodeType<S>>, T> :
  Y extends TRecursive<infer S> ? DecodeModifier<TRecursive<DecodeType<S>>, T> : 
  Y extends TRef<infer S> ? DecodeModifier<TRef<DecodeType<S>>, T> :
  Y extends TTuple<infer S> ? DecodeModifier<TTuple<AssertRest<DecodeRest<S>>>, T> : 
  Y extends TUnion<infer S> ? DecodeModifier<TUnion<AssertRest<DecodeRest<S>>>, T> : 
  T
```
If you want to experiment with the above during the week, that would be cool. I won't be able to spend much time with it until the weekend (possibly early next week), but if you can think of a way to simplify this type (which may involve investigating TOptional and TReadonly), can review later on and try get this patched on the next revision.

Cheers!

**Comment by sinclairzx81** [comment]
@HamishWHC Heya,

Had another look at this (thought I had a quick solution but still running into issues). The problem here may be a compiler bug but I'm not sure.

Can reproduce the issue using the following.

```typescript
const R1 = Type.Record(Type.Number(), Type.Boolean())
const R2 = Type.Optional(Type.Record(Type.Number(), Type.Boolean()))

type Resolve<T extends TSchema> = T extends TRecord<infer K, infer S> 
  ? [K, S] // expect [TNumber, TBoolean]
  : never

type T1 = Resolve<typeof R1> //  [TNumber, TBoolean]
type T2 = Resolve<typeof R2> //  [TSchema, TBoolean] ? why TSchema for the Key?
```
For some reason, TypeScript is able to resolve the `TNumber` key for the T1 result, but for the T2 result, the key is resolved as `TSchema` (however the value type is correct). TypeScript appears to be losing track of the key type through the optional wrap, and I'm currently somewhat at a loss to explain why.

This said, solving for the above T2 case would solve the StaticDecode inference case also. 

Documenting here for reference.

**Comment by sinclairzx81** [comment]
@HamishWHC Heya,

I have published a fix for this on `0.31.6`. Still somewhat baffled by the inference behavior above, but have opted for the recursive unwrap you had suggested on https://github.com/sinclairzx81/typebox/pull/544 (updated to include the unwrap for readonly). Had to make a few additional updates to the Value Decode and Encode functions (this to get around a TS instantiation quirk for overloaded functions), but with these amended, the original https://github.com/sinclairzx81/typebox/pull/544 unwrap seems ok to me.

> Also, the deep instantiation issue reported on https://github.com/sinclairzx81/typebox/issues/554 should be resolved in this update.

Overall, I think this StaticDecode type is probably at the upper limit of reasonable TS inference. TypeBox used to infer all types using this deep recursive approach but had shifted to a type caching implementation via the `static` property (as inference issues did pop up from time to time). I think implementing similar caching strategy for Transform types may be beneficial in the long run, but the work to implement this would be fairly vast. I'll keep an eye on the current implementation over the next few revisions (with a bit of luck the current implementation will hold until I've time to update to a caching approach)

Thanks again for reporting, will close off this issue for now.
All the best
S
